PR Enhanced Plugin Scheduler with Break/Pause System

Voxsylvae opened the pull request:
## Overview

This PR introduces a comprehensive overhaul of the Microbot plugin scheduler system, featuring advanced break and pause functionality, enhanced UI components, and a robust schedulable plugin API. The changes span across 4 major commits.

## Technical Summary

### Key Features Introduced

1. **Advanced Break and Pause System**

   - Seamless integration between PluginScheduler and BreakHandler
   - Plugin pause/resume functionality with PluginPauseEvent system
   - Intelligent break timing and duration management
   - Enhanced state management across scheduler and break operations

2. **Comprehensive Schedulable Plugin API**

   - Complete SchedulablePlugin interface with lifecycle hooks
   - Progress tracking and condition update mechanisms
   - Event-driven architecture for plugin management
   - Example implementation demonstrating full capabilities

3. **Enhanced User Interface**

   - Real-time scheduler info overlay for in-game status display
   - Comprehensive scheduler UI with table management
   - Dynamic layout components with improved responsiveness
   - Enhanced visual feedback and progress tracking

4. **Robust Condition System**

   - Improved logical conditions (LockCondition, PredicateCondition)
   - Enhanced resource and time condition evaluations
   - Better condition debugging and diagnostic capabilities
   - Advanced skill-based conditions with progress tracking

## Commit-by-Commit Analysis

### Commit 1: `e90c2f445e` - Final scheduler and breakhandler integration with UI enhancements

**Files Changed**: 24 files (+2,691 insertions, -661 deletions)

**Major Changes**:

- **UI Infrastructure**: Introduction of `DynamicFlowLayout` and `UIUtils` for responsive UI components
- **Enhanced Overlay System**: improvements to `SchedulerInfoOverlay` with real-time status display
- **Scheduler Logic Refinement**: Advanced plugin selection algorithms and improved state management
- **Resource Condition Enhancements**: Better regex pattern detection and randomization in resource panels
- **Time Condition Improvements**: Enhanced interval and time window condition handling
- **Utility Expansions**: Extended `SchedulerPluginUtil` with  plugin management capabilities

**Technical Highlights**:

- Implemented throttled UI updates to prevent excessive refresh cycles
- Added plugin prioritization algorithms
- Enhanced break integration with extend/interrupt capabilities
- Improved condition evaluation performance and accuracy

### Commit 2: `679b545797` - Add scheduler info overlay and improve plugin scheduling logic

**Files Changed**: 15 files (+1,327 insertions, -424 deletions)

**Major Changes**:

- **SchedulerInfoOverlay Introduction**: New overlay component providing real-time scheduler status
- **Enhanced Plugin Selection**: Improved algorithms for plugin selection and prioritization
- **UI/UX Improvements**: Better visual feedback and user experience enhancements
- **API Enhancements**: Extended schedulable plugin capabilities
- **Condition System Refinements**: Better condition evaluation and progress tracking

**Technical Highlights**:

- Real-time status display with minimal performance impact
- Improved plugin lifecycle management
- Enhanced error handling and recovery mechanisms
- Better integration with existing RuneLite overlay system

### Commit 3: `41adb91138` - Enhanced BreakHandler with scheduler integration and pause system

**Files Changed**: 5 files (+391 insertions, -120 deletions)

**Major Changes**:

- **BreakHandler Overhaul**: Complete restructuring of `BreakHandlerScript` (460+ lines of improvements)
- **Pause System Integration**: Deep integration with PluginPauseEvent system
- **Enhanced Configuration**: Extended break behavior configuration options
- **State Management**: Improved break state tracking and validation
- **Timing Improvements**: Better break timing calculations and duration management

**Technical Highlights**:

- Robust break scheduling with intelligent timing algorithms
- Seamless coordination with scheduler pause operations
- Enhanced plugin lifecycle integration during breaks
- Improved break interruption and continuation handling
- Better visual feedback during break operations

### Commit 4: `ea35084c55` - Enhanced plugin scheduler with comprehensive UI, pause system and schedulable API

**Files Changed**: 62 files (+6,654 insertions, -1,426 deletions)

**Major Changes**:

- **Core Scheduler Transformation**: Complete overhaul of `SchedulerPlugin` with advanced pause/resume functionality
- **SchedulablePlugin API**: Comprehensive interface with full lifecycle hooks and progress tracking
- **UI Framework**: Complete scheduler UI with table management and real-time updates
- **Condition System Expansion**: Enhanced logical, resource, and time conditions
- **Documentation**: Extensive documentation including user guides and API references
- **Event System**: Introduction of `PluginPauseEvent` for coordinated operations

**Technical Highlights**:

- Advanced pause/resume functionality for running plugins
- Plugin manual start functionality with comprehensive safety checks
- Enhanced state management with comprehensive scheduler states
- Complete UI framework with form panels and configuration management
- condition evaluation system with debugging capabilities

## Architecture Improvements

### Scheduler Core

- **State Management**: Comprehensive `SchedulerState` enum covering all operational states
- **Plugin Lifecycle**: Enhanced lifecycle management with proper startup, pause, resume, and shutdown sequences
- **Break Integration**: Seamless integration between scheduling and break operations
- **Error Handling**: Robust error handling with proper recovery mechanisms

### Break System

- **Intelligent Timing**: Advanced algorithms for break duration and frequency calculations
- **State Coordination**: Proper coordination between scheduler and break states
- **Configuration Flexibility**: Extensive configuration options for break behavior customization
- **Visual Feedback**: Enhanced overlay system providing clear break status information

### UI Framework

- **Responsive Design**: Dynamic layout components that adapt to content
- **Real-time Updates**: Efficient update mechanisms with proper throttling
- **User Experience**: Intuitive interfaces for complex scheduling configurations
- **Visual Consistency**: Consistent styling and behavior across all UI components

### API Design

- **SchedulablePlugin Interface**: Comprehensive interface enabling full plugin lifecycle management
- **Event System**: Event-driven architecture supporting coordinated operations
- **Progress Tracking**: Built-in progress tracking and reporting mechanisms
- **Extensibility**: Designed for easy extension and customization


## Documentation

### User Documentation

- Comprehensive user guide for scheduler functionality
- Plugin writer's guide for implementing schedulable plugins
- Condition system documentation with examples
- API reference documentation

## Compatibility

### Backward Compatibility

- Existing plugins continue to work without modification
- Gradual migration path for plugins to adopt new features
- Legacy condition system still supported
- Configuration migration handled automatically

**Total Changes**: 106 files modified, 10,000+ lines added/modified  
**Primary Focus Areas**: Plugin Scheduler, Break Handler, UI Framework, API Design

Voxsylvae committed:
feat: Enhanced plugin scheduler with comprehensive UI, pause system and schedulable API

- Advanced pause/resume functionality for running plugins
- Improved plugin scheduling logic with priority handling
- Enhanced state management with comprehensive scheduler states
- Plugin manual start functionality with safety checks
- Break integration with extend/interrupt capabilities

- Complete SchedulablePlugin interface with lifecycle hooks
- Plugin pause/resume integration with PluginPauseEvent
- Progress tracking and condition update mechanisms
- Example plugin demonstrating full schedulable implementation

- Comprehensive plugin scheduler UI with table management
- Real-time status updates and progress tracking
- Enhanced info panel with break timing and status
- Improved form panels for plugin configuration

- Logical condition improvements (LockCondition, SkillLevelCondition)
- Enhanced condition evaluation and progress tracking
- Better condition debugging and diagnostic capabilities

- PluginPauseEvent for coordinated pause/resume operations
- Enhanced plugin lifecycle event handling
- Improved event-driven architecture for plugin management

- Extended SchedulerPluginUtil with break management
- Enhanced plugin sorting and selection algorithms
- Improved plugin state validation and error handling

This commit provides a comprehensive foundation for advanced plugin scheduling
with full pause/resume capabilities, enhanced UI, and robust condition system.

Voxsylvae committed:
feat: Enhanced BreakHandler with scheduler integration and pause system

- Integrated pause/resume functionality with PluginPauseEvent system
- Enhanced break timing and duration management
- Improved break state tracking and validation
- Better integration with scheduler pause/resume operations

- Extended configuration options for break behavior
- Enhanced break timing controls and customization
- Improved break scheduling parameters

- Better coordination with PluginScheduler pause system
- Enhanced plugin lifecycle integration during breaks
- Improved break interruption and continuation handling

- Comprehensive BreakHandlerScript improvements (460+ lines)
- Enhanced break logic with better state management
- Improved break timing calculations and validation
- Better error handling and recovery mechanisms

- Enhanced break status display and information
- Better visual feedback during break operations
- Improved overlay positioning and styling

This commit provides robust break handling capabilities that integrate
seamlessly with the enhanced plugin scheduler pause system.

Voxsylvae committed:
feat: Add scheduler info overlay and improve plugin scheduling logic

- Add new SchedulerInfoOverlay with status icons and timing display
- Enhance plugin selection algorithm with better prioritization
- Add UI update throttling to prevent excessive refreshes
- Improve resource condition panel with automatic regex detection
- Refactor method names for better clarity and consistency
- Add foundation for scheduler group functionality

Voxsylvae committed:
feat: Final scheduler and breakhandler integration with UI enhancements

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/Microbot.java:
```diff
@@ -419,10 +420,13 @@ public static List<Rs2ItemModel> updateItemContainer(int id, ItemContainerChange
+        // Check if we're currently on the client thread for logging purposes
+        boolean isOnClientThread = getClient().isClientThread();                
+        // This makes the method non-blocking regardless of which thread it's called from		
+        if(isOnClientThread) {
```
Possibly this would be a better option? considering runOnClientThreadOptional does consider if the method is being ran on the client thread or not.

```java
@SneakyThrows
private static boolean togglePlugin(Plugin plugin, boolean enable) {
	if (plugin == null) return false; // we should always be returning false, e.g - if enable is false and plugin is null - !enable, would return true

	final AtomicBoolean success = new AtomicBoolean(false);

	Callable<Boolean> callable = () -> {
		try {
			getPluginManager().setPluginEnabled(plugin, enable);
			if (enable) {
				success.set(getPluginManager().startPlugin(plugin));
				getPluginManager().startPlugins();
			} else {
				success.set(getPluginManager().stopPlugin(plugin));
			}
		} catch (PluginInstantiationException e) {
			log.error("Error toggling plugin ({}): {}", plugin.getClass().getSimpleName(), e.getMessage(), e);
		}
		return success.get();
	};

	return getClientThread().runOnClientThreadOptional(callable).orElse(false);
}
```

Voxsylvae replied:
Good Suggestion, i forgot about runOnClientThreadOptional already checks for if it runs on the game client thread, Sry.

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/Microbot.java:
```diff
@@ -492,6 +515,35 @@ public static boolean stopPlugin(Class<? extends Plugin> clazz) {
+     *
+     * @param plugin the plugin to be stopped.
+     */
+    public static void stopPluginV(Plugin plugin) {
```
what would the purpose of this method?

Voxsylvae replied:
Oh Sry forget to delete it, on a merge...

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/BlockingEventManager.java:
```diff
@@ -52,7 +52,7 @@ public BlockingEventManager()
         blockingEvents.add(new BankTutorialEvent());
         blockingEvents.add(new DeathEvent());
         blockingEvents.add(new BankJagexPopupEvent());
-        blockingEvents.add(new ScriptPauseEvent()); // Add this line
```
Possibly we should remove script pause event ?

gmason0 reviewed (changes requested):
Looks great! just a bit of TLC then we should be good to go.

Voxsylvae committed:
feat: improve plugin scheduler stop logic and state management

Major Improvements:
- Fixed scheduler soft stop mechanism with proper @Subscribe annotation usage
- Enhanced plugin stop state management with better monitoring thread handling
- Improved GiantSeaweedFarmer plugin with lock conditions and state-aware stopping
- Refined plugin toggle mechanism to use consistent client thread execution

Stop Feature Enhancements:
- Renamed isHardStoppable() to allowHardStop() for better API clarity
- Enhanced stop monitoring with better logging and state tracking
- Fixed stop condition evaluation and retry logic
- Added cancelStop() functionality for better stop control
- Improved timing validation with proper randomization unit handling

Bug Fixes:
- Fixed overlay position and hiding logic for better UI experience
- Corrected TimeWindowCondition randomization calculations and limits
- Fixed SchedulablePlugin event handlers to use proper EventBus subscription
- Removed unused ScriptPauseEvent class
- Enhanced error handling in plugin stopping with better timeouts

State Logic Improvements:
- Better state transition handling in plugin lifecycle
- Enhanced condition management with proper lock/unlock patterns
- Improved plugin finish reporting with proper thread execution
- Fixed break handler lock state logging and management

Documentation Updates:
- Updated API documentation to reflect method name changes
- Enhanced inline documentation for stop mechanisms
- Improved tooltip information for time condition UI elements

gmason0 reviewed (approved):
Thanks again for the amazing improvements & hard work, keep it up :)
